firebase-firestore/src/main/java/com/google/firebase/firestore/core/RelationFilter.java

```diff
@@ -74,6 +77,17 @@ public boolean matches(Document doc) {
   private boolean matchesValue(FieldValue other) {
     if (operator == Operator.ARRAY_CONTAINS) {
       return other instanceof ArrayValue && ((ArrayValue) other).getInternalValue().contains(value);
+    } else if (operator == Operator.IN) {
+      return value instanceof ArrayValue && ((ArrayValue) value).getInternalValue().contains(other);
+    } else if (operator == Operator.ARRAY_CONTAINS_ANY) {
+      if (other instanceof ArrayValue && value instanceof ArrayValue) {
```

Comment:
this `instanceof` check on `value` should be an assert as well (but not for `other` since `other` is the document field value and may or may not be an array).